### PR TITLE
Handle Arithmetic operations for fields which are in network order

### PR DESCRIPTION
**Pull request by komaljai**
Handle Arithmetic operations for fields which are in network order(annotated in header definition, action parameters and table keys)

### CODE REVIEW

File: backends/ebpf/target.h [hunk 1]
```diff
@@ -199,6 +201,49 @@ class KernelSamplesTarget : public Target {
 class P4TCTarget : public KernelSamplesTarget {
  public:
     explicit P4TCTarget(bool emitTrace) : KernelSamplesTarget(emitTrace, "P4TC") {}
+    bool isAnnotated(const IR::Vector<IR::Annotation> annotations) const {
```
Comment (komaljai) [inline backends/ebpf/target.h:204]: change function name

Comment (komaljai) [reply backends/ebpf/target.h]: addressed

File: testdata/p4tc_samples_outputs/ipip_control_blocks.c [hunk 2]
```diff
@@ -182,7 +182,6 @@ if (/* hdr->outer.isValid() */
                 return TC_ACT_SHOT;
             }
             
-            hdr->ethernet.dstAddr = htonll(hdr->ethernet.dstAddr << 16);
```
Comment (vbnogueira) [inline testdata/p4tc_samples_outputs/ipip_control_blocks.c:185]: I found an issue which is not shown in any example you have here, however if you remove the `@tc_type("macaddr")` annotation from the ipip ethernet header (https://github.com/komaljai/p4c/blob/networkOrder_arithmetic/testdata/p4tc_samples/ipip.p4#L11) you'll see it

When it converts the mac address to host endian (note that in this case there will be no annotation in the header) https://github.com/komaljai/p4c/blob/networkOrder_arithmetic/testdata/p4tc_samples_outputs/ipip_control_blocks.c#L104, it will do this:

```
hdr->ethernet.dstAddr = ntohll(value->u.Main_set_nh.dmac);
```

However the deparser will try to convert back shifting, as in the line above:

```
hdr->ethernet.dstAddr = htonll(hdr->ethernet.dstAddr << 16);
```

So the mac address will be written in the deparser skipping a couple bytes

File: testdata/p4tc_samples_outputs/checksum_parser.c [hunk 3]
```diff
@@ -106,7 +106,7 @@ static __always_inline int run_parser(struct __sk_buff *skb, struct my_ingress_h
 ;
             tmp_0 = /* ck_0.get() */
 ((u16) (~ck_0_state));/* verify(hdr->ipv4.hdrChecksum == tmp_0, BadIPv4HeaderChecksum) */
-            if (!(hdr->ipv4.hdrChecksum == tmp_0)) {
+            if (!((hdr->ipv4.hdrChecksum == tmp_0))) {
```
Comment (usha1830) [inline testdata/p4tc_samples_outputs/checksum_parser.c:109]: extra parentheses pair

File: backends/ebpf/codeGen.cpp [hunk 4]
```diff
@@ -458,6 +474,141 @@ void CodeGenInspector::widthCheck(const IR::Node *node) const {
...
+    } else if (widthToEmit <= 64) {
+        emit = byte_order == "HOST" ? "ntohll" : "bpf_cpu_to_be64";
+    }
+    builder->appendFormat("%s(", emit);
+    visit(expr);
+    builder->append(")");
+}
+
+void CodeGenInspector::emitTCBinaryOperation(const IR::Operation_Binary *b, bool isScalar) {
+    if (b->is<IR::Operation_Binary>()) {
```
Comment (usha1830) [inline backends/ebpf/codeGen.cpp:495]: This is already coming as IR::Operation_Binary *b, why is this necessary?

**Review by maheswari-s — APPROVED** [review]
Looks fine

File: backends/ebpf/codeGen.cpp [hunk 5]
```diff
@@ -458,6 +474,141 @@ void CodeGenInspector::widthCheck(const IR::Node *node) const {
...
+        unsigned width = dynamic_cast<IHasWidth *>(et)->widthInBits();
+        visit(lexpr);
+        if (isScalar) {
+            builder->spc();
+            builder->append(stringop);
+            builder->spc();
+        } else {
+            builder->append(", &");
+        }
+        expressionPrecedence = b->getPrecedence() + 1;
```
Comment (usha1830) [inline backends/ebpf/codeGen.cpp:557]: What does this statement do? expressionPrecedence looks like global, would it impact ebpf backend in any way?

**Review by usha1830 — APPROVED** [review]
Looks good. Have a few comments/queries. Please check

File: testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c [hunk 6]
```diff
@@ -113,8 +113,8 @@ if (/* hdr->ipv4.isValid() */
                                 break;
                             case MAINCONTROLIMPL_IPV4_TBL_1_ACT_MAINCONTROLIMPL_SEND_NH: 
                                 {
-                                    hdr->ethernet.srcAddr = value->u.MainControlImpl_send_nh.smac;
-                                                                        hdr->ethernet.dstAddr = value->u.MainControlImpl_send_nh.dmac;
+                                    hdr->ethernet.srcAddr = bpf_cpu_to_be64(value->u.MainControlImpl_send_nh.smac);
+                                                                        hdr->ethernet.dstAddr = bpf_cpu_to_be64(value->u.MainControlImpl_send_nh.dmac);
```
[the remark this answers is not in the document]
Comment (komaljai) [reply testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c:117]: This is intentional, we wanted to check scenario if it is not annotated, will compiler take care of version.

[the remark this answers is not in the document]
Comment (vbnogueira) [reply testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c:117]: I see, so this is like an error case?
I understand that it will compile find (so it wouldn't be considered an error case by the compiler), but it would not work properly in the datapath

[the remark this answers is not in the document]
Comment (komaljai) [reply testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c:117]: Yes it will be compiled properly but will be error case in datapath.

**Review by usha1830 — APPROVED** [review]
